File: replace-plugin/src/main/java/com/maple/replace/replace/ReplaceClassVisitor.java
package com.maple.replace.replace;

import com.android.ddmlib.Log;

import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;

import java.util.List;

/**
 * 代码行替换
 */
public class ReplaceClassVisitor extends ClassVisitor implements Opcodes {
    private String mClassName;
    private final List<ReplaceBean> replaceBeans = ReplaceConfig.getReplaceBeans();

    public ReplaceClassVisitor(ClassVisitor cv) {
        super(Opcodes.ASM5, cv);
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        // 11:12:25 E/MS_ASM: visit:51 - 33
        // - androidx/fragment/app/FragmentActivity
        // - null
        // - androidx/activity/ComponentActivity
//        Log.e("MS_ASM", "visit:" + version + " - " + access
//                + " - " + name
//                + " - " + signature
//                + " - " + superName
//        );
        this.mClassName = name;
        super.visit(version, access, name, signature, superName, interfaces);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        // 11:56:27 E/MS_ASM: visitMethod:4 - onCreate - (Landroid/os/Bundle;)V - null
        // Log.e("MS_ASM", "visitMethod:" + access + " - " + name + " - " + desc + " - " + signature);
        ReplaceBean targetBean = null;
        if (replaceBeans.size() > 0) {
            for (ReplaceBean bean : replaceBeans) {
                if (mClassName.equals(bean.getNewOwner()) && name.equals(bean.getNewName())) {
                    targetBean = bean;
                }
            }
        }

        MethodVisitor mv = cv.visitMethod(access, name, desc, signature, exceptions);
        if (targetBean != null) {
            Log.e("MS_ASM", "- 跳过原方法:" + targetBean.getNewOwner() + " - " + targetBean.getNewName());
            return mv;
        } else {
            return new ReplaceMethodVisitor(mv, replaceBeans);
        }
    }

}

File: base-plugin/src/main/kotlin/com/maple/plugin/plugins/replace/ReplaceClassVisitor.kt
package com.maple.plugin.plugins.replace

import com.android.ddmlib.Log
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

/**
 * 代码行替换
 */
class ReplaceClassVisitor(
    cv: ClassVisitor,
    val configs: List<ReplaceBean>
) : ClassVisitor(Opcodes.ASM5, cv), Opcodes {
    private var mClassName: String = ""

    override fun visit(
        version: Int, // 51
        access: Int,  // 33
        name: String, // androidx/fragment/app/FragmentActivity
        signature: String?,
        superName: String, // androidx/activity/ComponentActivity
        interfaces: Array<String>?
    ) {
        mClassName = name
        super.visit(version, access, name, signature, superName, interfaces)
    }

    override fun visitMethod(
        access: Int,  // 4
        name: String, // onCreate
        desc: String, // (Landroid/os/Bundle;)V
        signature: String?,
        exceptions: Array<String>?
    ): MethodVisitor {
        // 4 - onCreate - (Landroid/os/Bundle;)V - null
        // Log.e("MS_ASM", "visitMethod:$access - $name - $desc - $signature")
        val replace = configs.find {
            mClassName == it.newOwner && name == it.newName
        }
        val vm = super.visitMethod(access, name, desc, signature, exceptions)
        return if (replace != null) {
            Log.e("MS_ASM", "跳过原方法: ${replace.newOwner} - ${replace.newName}")
            vm
        } else {
            getReplaceMethodVisitor(vm)
        }
    }

    /**
     * 替换方法体中某一行代码
     */
    private fun getReplaceMethodVisitor(vm: MethodVisitor) = object : MethodVisitor(Opcodes.ASM5, vm) {
        override fun visitMethodInsn(opcode: Int, owner: String?, name: String?, descriptor: String?, isInterface: Boolean) {
            val bean = configs.find {
                it.isSameOldConfig(opcode, owner, name, descriptor, isInterface)
            }
            if (bean != null) {
                // ctx.sendBroadcast(intent);
                // 182 - android/content/Context - sendBroadcast - (Landroid/content/Intent;)V - false

                // BroadcastUtils.sendAppInsideBroadcast(ctx, intent);
                // 184 - com/gavin/asmdemo/BroadcastUtils - sendAppInsideBroadcast - (Landroid/content/Context;Landroid/content/Intent;)V - false
                Log.e("MS_ASM", "执行替换: $opcode - $owner - $name - $descriptor - $isInterface")
                super.visitMethodInsn(bean.getNewOpcodeInt(), bean.newOwner, bean.newName, bean.newDescriptor, bean.newIsInterface)
            } else {
                super.visitMethodInsn(opcode, owner, name, descriptor, isInterface)
            }
        }
    }

}